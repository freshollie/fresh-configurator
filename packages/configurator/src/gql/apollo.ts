/**
 * Patch apollo to remove pollIntervals from useQuery when
 * skip is set: https://github.com/apollographql/react-apollo/issues/2127
 */

import {
  ApolloClient,
  InMemoryCache,
  QueryResult,
  useQuery as useQueryOrig,
  QueryHookOptions,
  OperationVariables,
  TypedDocumentNode,
} from "@apollo/client";
import { DocumentNode } from "graphql";

export * from "@apollo/client";

// We are following the definitions from apollo, so any has to be used here
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useQuery<TData = any, TVariables = OperationVariables>(
  query: DocumentNode | TypedDocumentNode<TData, TVariables>,
  options?: QueryHookOptions<TData, TVariables>
): QueryResult<TData, TVariables> {
  return useQueryOrig(
    query,
    options
      ? {
          ...options,
          pollInterval:
            options.skip && options.pollInterval ? 0 : options.pollInterval,

          // IMPORTANT: Apollo has some bug where components don't rerender when
          // they should unless this is enabled!
          notifyOnNetworkStatusChange: !options.pollInterval,
        }
      : undefined
  );
}

export type ApolloContext = { client: ApolloClient<InMemoryCache> };
