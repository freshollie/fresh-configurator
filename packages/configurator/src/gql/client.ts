import { InMemoryCache, ApolloClient, gql } from "@apollo/client";
import WebSocketLink from "./WebSocketLink";
import { Resolvers, Configurator } from "./__generated__";
import introspection from "./__generated__/introspection.json";
import { versionInfo } from "../util";
import {
  LogsDocument,
  SelectedTabDocument,
  ConnectionSettingsDocument,
} from "./queries/Configurator.graphql";

const typeDefs = gql`
  type Query {
    configurator: Configurator!
  }

  type Mutation {
    setConnectionSettings(port: String!, baudRate: Int): Boolean
    setConnection(connection: ID): Boolean
    setConnecting(value: Boolean!): Boolean
    setTab(tabId: String!): Boolean
    setExpertMode(enabled: Boolean!): Boolean
    log(message: String!): Boolean
  }

  type Configurator {
    port: String
    connection: ID
    connecting: Boolean!
    baudRate: Int!
    tab: String
    expertMode: Boolean!
    logs: [Log!]!
  }

  type Log {
    time: String!
    message: String!
  }
`;

export const cache = (): InMemoryCache =>
  new InMemoryCache({
    typePolicies: Object.fromEntries(
      // eslint-disable-next-line no-underscore-dangle
      introspection.__schema.types
        .filter(({ kind }) => kind === "OBJECT")
        .map(({ name }) => [name, { merge: true }])
    ),
  });

export const resolvers = (initialState?: {
  connecting?: boolean;
  connection?: string | null;
  port?: string;
  baudRate?: number;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}): any => {
  const config: Resolvers = {
    Query: {
      configurator: () => ({
        __typename: "Configurator",

        port: "",
        baudRate: 115200,
        connecting: false,
        connection: null,
        expertMode: false,
        logs: [],
        ...initialState,
      }),
    },
    Mutation: {
      setTab: (_, { tabId }, { client }) => {
        client.writeQuery({
          query: SelectedTabDocument,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              tab: tabId,
            },
          },
        });
        return null;
      },
      setConnectionSettings: (_, { port, baudRate }, { client }) => {
        const existingData = {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          ...client.readQuery({
            query: ConnectionSettingsDocument,
          })!.configurator,
        };

        if (typeof baudRate === "number") {
          existingData.baudRate = baudRate;
        }
        existingData.port = port;

        client.writeQuery({
          query: ConnectionSettingsDocument,
          data: {
            __typename: "Query",
            configurator: {
              ...existingData,
            },
          },
        });

        return null;
      },
      setConnecting: (_, { value }, { client }) => {
        client.writeQuery({
          query: gql`
            query {
              configurator {
                connecting
              }
            }
          `,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              connecting: value,
            },
          },
        });

        return null;
      },
      setConnection: (_, { connection }, { client }) => {
        client.writeQuery({
          query: gql`
            query {
              configurator {
                connection
              }
            }
          `,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              connection,
            },
          },
        });

        return null;
      },
      // setExpertMode: (_, { enabled }, { client }) => {
      //   client.writeQuery({
      //     document: Ex,
      //     data: {
      //       __typename: "Query",
      //       configurator: {
      //         __typename: "Configurator",
      //         expertMode: enabled,
      //       } as Configurator,
      //     },
      //   });

      //   return null;
      // },
      log: (_, { message }, { client }) => {
        const logs =
          client.readQuery({
            query: LogsDocument,
          })?.configurator.logs ?? [];

        client.writeQuery({
          query: LogsDocument,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              logs: logs.concat([
                {
                  time: new Date().toISOString(),
                  message,
                  __typename: "Log" as const,
                },
              ]),
            },
          },
        });

        return null;
      },
    },
  };
  return config;
};

// extract the backend address from the URL search query, as this can
// be dynamically passed to us by electron
const searchParams = new URLSearchParams(window.location.search.slice(1));
const BACKEND = searchParams.get("backend") ?? "ws://localhost:9000";

const client = new ApolloClient({
  cache: cache(),
  typeDefs,
  resolvers: resolvers(),
  link: new WebSocketLink({
    url: `${BACKEND}/graphql`,
    keepAlive: 99999999999,
  }),
});

const writeInitial = (): void => {
  const { os, version, chromeVersion } = versionInfo();
  client.writeQuery({
    query: gql`
      query {
        configurator {
          port
          baudRate
          tab
          expertMode
          connecting
          connection
          logs {
            time
            message
          }
        }
      }
    `,
    data: {
      __typename: "Query",
      configurator: {
        port: null,
        baudRate: 115200,
        tab: null,
        expertMode: false,
        connecting: false,
        connection: null,
        logs: [
          {
            time: new Date().toISOString(),
            message: `Running - OS: <strong>${os}</strong>, Chrome: <strong>${chromeVersion}</strong>, Configurator: <strong>${version}</strong>`,
            __typename: "Log",
          },
        ],
        __typename: "Configurator",
      } as Configurator,
    },
  });
};

writeInitial();
client.onResetStore(() => Promise.resolve(writeInitial()));

export default client;
