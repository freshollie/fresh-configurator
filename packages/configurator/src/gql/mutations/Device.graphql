mutation CallibrateAccelerometer($connection: ID!) {
  deviceCallibrateAccelerometer(connectionId: $connection)
}

mutation SetArming(
  $connection: ID!
  $armingDisabled: Boolean!
  $runawayTakeoffPreventionDisabled: Boolean!
) {
  deviceSetArming(
    connectionId: $connection
    armingDisabled: $armingDisabled
    runawayTakeoffPreventionDisabled: $runawayTakeoffPreventionDisabled
  )
}

mutation Reset($connection: ID!) {
  deviceReset(connectionId: $connection)
}

mutation SetBoardAlignment($connection: ID!, $alignment: AlignmentInput!) {
  deviceSetBoardAlignment(connectionId: $connection, alignment: $alignment)
}

mutation SetPidProtocols($connection: ID!, $protocols: PidProtocolsInput!) {
  deviceSetPidProtocols(connectionId: $connection, protocols: $protocols)
}

mutation SetMotorsDirection($connection: ID!, $reversed: Boolean!) {
  deviceSetMotorsDirection(connectionId: $connection, reversed: $reversed)
}

mutation SetMotorDigitalIdleSpeed($connection: ID!, $idlePercentage: Float!) {
  deviceSetDigitalIdleSpeed(
    connectionId: $connection
    idlePercentage: $idlePercentage
  )
}

mutation SetDshotBeeperConfig(
  $connection: ID!
  $config: DshotBeeperConfigInput!
) {
  deviceSetDshotBeeperConfig(connectionId: $connection, config: $config)
}

mutation SetDisabledSensors($connection: ID!, $disabledSensors: [Int!]!) {
  deviceSetDisabledSensors(
    connectionId: $connection
    disabledSensors: $disabledSensors
  )
}
