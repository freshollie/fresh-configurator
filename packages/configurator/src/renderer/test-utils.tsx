import React from "react";
import { render as renderFunction, RenderResult } from "@testing-library/react";
import { Provider } from "bumbag";
import theme from "./theme";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: (query: string) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  }),
});

const AllTheProviders: React.FC = ({ children }) => (
  <Provider theme={theme}>{children}</Provider>
);

const customRender = (
  ui: Parameters<typeof renderFunction>[0],
  options?: Parameters<typeof renderFunction>[1]
): RenderResult => renderFunction(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
