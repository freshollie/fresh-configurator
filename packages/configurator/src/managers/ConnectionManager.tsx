import React, { useEffect, useCallback } from "react";
import semver from "semver";
import { OnConnectionChangedDocument } from "../gql/queries/Connection.graphql";
import config from "../config";
import Icon from "../components/Icon";
import useConnectionState from "../hooks/useConnectionState";
import useLogger from "../hooks/useLogger";
import { ConnectionSettingsDocument } from "../gql/queries/Configurator.graphql";
import {
  ConnectDocument,
  DisconnectDocument,
} from "../gql/mutations/Connection.graphql";
import { SetArmingDocument } from "../gql/mutations/Device.graphql";

import BigButton from "../components/BigButton";
import { useMutation, useQuery, useSubscription } from "../gql/apollo";

/**
 * Handle all aspects of tracking the app connection
 * to the device, as well as providing an interface
 * for the user to "connect" or "disconnect"
 */
const ConnectionManager: React.FC = () => {
  const log = useLogger();

  const { data: configuratorQuery } = useQuery(ConnectionSettingsDocument);
  const { port, baudRate } = configuratorQuery?.configurator ?? {};

  const {
    connecting,
    connection,
    setConnection,
    setConnecting,
  } = useConnectionState();

  const [disableArming] = useMutation(SetArmingDocument, {
    onCompleted: () => {
      log("<b>Arming disabled</b>");
    },
    onError: (e) => {
      log(
        `<span class="message-negative">Error disabling arming: ${e.message}</span>`
      );
    },
  });

  const [disconnectMutation] = useMutation(DisconnectDocument);

  const disconnect = (
    notify = true,
    connectionId = connection
  ): Promise<void> =>
    disconnectMutation({
      variables: {
        connection: connectionId ?? "",
      },
    })
      .then(() => {
        if (notify) {
          log(
            `Serial port <span class="message-positive">successfully</span> closed for connectionId=${connectionId}`
          );
        }
      })
      .catch((e) => {
        if (notify) {
          log(`Error closing connection: ${e.message}`);
        }
      });

  const [connect] = useMutation(ConnectDocument, {
    variables: {
      port: port ?? "",
      baudRate: baudRate ?? 0,
    },
    onCompleted: ({ connect: { id: connectionId, apiVersion } }) => {
      if (!connecting) {
        disconnect(false, connectionId);
        return;
      }

      log(
        `Serial port <span class="message-positive">successfully</span> opened, connectionId=${connectionId}`
      );
      log(`MultiWii API version: <strong>${apiVersion}</strong>`);
      if (semver.lt(apiVersion, config.apiVersionAccepted)) {
        log(
          `MSP version not supported: <span class="message-negative">${apiVersion}</span>`
        );
        disconnect(true, connectionId);
      } else {
        setConnection(connectionId);
        disableArming({
          variables: {
            connection: connectionId,
            armingDisabled: true,
            runawayTakeoffPreventionDisabled: false,
          },
        });
      }
      setConnecting(false);
    },
    onError: (e) => {
      if (!connecting) {
        return;
      }
      log(
        `Could not open connection (<span class="message-negative">${e.message}</span>), communication <span class="message-negative">failed</span>`
      );
      setConnecting(false);
    },
  });

  const connectionClosed = useCallback(() => {
    setConnection(null);
    log(`Serial port closed unexpectedly for connectionId=${connection}`);
  }, [connection, log, setConnection]);

  // Create a subscription to the current connection
  // and handle updating the app state when connection
  // is closed
  const { error: subscriptionError } = useSubscription(
    OnConnectionChangedDocument,
    {
      variables: {
        connection: connection ?? "",
      },
      skip: !connection,
      onSubscriptionData: ({ subscriptionData: { data: onChangedData } }) => {
        const newConnectionId = onChangedData?.onConnectionChanged;
        if (newConnectionId) {
          setConnection(newConnectionId);
          log(
            `Serial port reconnected for connectionId=${connection} newConnectionId=${newConnectionId} `
          );
        } else if (newConnectionId === null) {
          connectionClosed();
        }
      },
    }
  );

  useEffect(() => {
    if (subscriptionError) {
      connectionClosed();
    }
  }, [connectionClosed, log, subscriptionError]);

  const handleClicked = (): void => {
    if (!connection && !connecting) {
      log(`Opening connection on ${port}`);
      setConnecting(true);
      connect();
    } else {
      if (connection) {
        disconnect();
        setConnection(null);
      } else if (connecting) {
        log(`Connection attempt to ${port} aborted`);
      }
      setConnecting(false);
    }
  };

  let statusText = "Connecting";

  if (!connecting) {
    statusText = connection ? "Disconnect" : "Connect";
  }

  const disconnected = !connection && !connecting;

  return (
    <BigButton
      active={!!connection}
      data-testid={disconnected ? "connect-button" : "disconnect-button"}
      icon={<Icon name={disconnected ? "usb-connect" : "usb-disconnect"} />}
      text={statusText}
      onClick={handleClicked}
    />
  );
};

export default ConnectionManager;
