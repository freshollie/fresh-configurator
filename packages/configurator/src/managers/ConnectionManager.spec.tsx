import React from "react";
import {
  MockedProvider,
  MockedResponse,
  MockSubscriptionLink,
} from "@apollo/client/testing";
import { GraphQLError } from "graphql";
import { InMemoryCache } from "@apollo/client";
import { render, waitFor, fireEvent } from "../test-utils";
import {
  ConnectDocument,
  DisconnectDocument,
} from "../gql/mutations/Connection.graphql";
import { OnConnectionChangedDocument } from "../gql/queries/Connection.graphql";
import {
  ConnectionStateDocument,
  LogsDocument,
} from "../gql/queries/Configurator.graphql";
import { SetArmingDocument } from "../gql/mutations/Device.graphql";
import { Connection } from "../gql/__generated__";
import ConnectionManager from "./ConnectionManager";
import { cache as clientCache, resolvers } from "../gql/client";

const connectMock = (
  port: string,
  baudRate: number,
  connectionId: string,
  apiVersion: string,
  onConnect = () => {}
): MockedResponse => ({
  request: {
    query: ConnectDocument,
    variables: {
      port,
      baudRate,
    },
  },
  result: () => {
    onConnect();

    return {
      data: {
        connect: {
          id: connectionId,
          apiVersion,
          __typename: "Connection",
        } as Connection,
      },
    };
  },
  delay: 100,
});

const connectErrorMock = (port: string, baudRate: number): MockedResponse => ({
  request: {
    query: ConnectDocument,
    variables: {
      port,
      baudRate,
    },
  },
  error: new Error("some error"),
  delay: 100,
});

const disconnectMock = (
  connectionId: string,
  onDisconnect: () => void
): MockedResponse => ({
  request: {
    query: DisconnectDocument,
    variables: {
      connection: connectionId,
    },
  },
  result: () => {
    onDisconnect();
    return {
      data: {
        close: connectionId,
      },
    };
  },
});

const setArmingMock = (
  connection: string,
  armingDisabled: boolean,
  runawayTakeoffPreventionDisabled: boolean,
  onSet: () => void
): MockedResponse => ({
  request: {
    query: SetArmingDocument,
    variables: {
      connection,
      armingDisabled,
      runawayTakeoffPreventionDisabled,
    },
  },
  result: () => {
    onSet();
    return {
      data: {
        deviceSetArming: null,
      },
    };
  },
});

const onChangedMock = (
  connectionId: string,
  response: string | null,
  onSubscribed = () => {}
): MockedResponse => ({
  request: {
    query: OnConnectionChangedDocument,
    variables: {
      connection: connectionId,
    },
  },
  delay: 99999,
  result: () => {
    onSubscribed();
    return {
      data: {
        onConnectionChanged: response,
      },
    };
  },
});

const connectionState = (
  cache: InMemoryCache
): { connecting: boolean; connection?: string | null } =>
  cache.readQuery({
    query: ConnectionStateDocument,
  })!.configurator;

const logs = (cache: InMemoryCache) =>
  cache
    .readQuery({ query: LogsDocument })!
    .configurator.logs.map((line) => line.message);

describe("ConnectionManager", () => {
  it("should connect with the connection settings and set the new connection id when connect clicked", async () => {
    const connect = jest.fn();
    const setDisarmed = jest.fn();
    const mockConnectionId = "23455";

    const cache = clientCache();
    const { getByTestId, asFragment } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: null,
          port: "/dev/something",
          baudRate: 115200,
        })}
        mocks={[
          connectMock(
            "/dev/something",
            115200,
            mockConnectionId,
            "1.40.1",
            connect
          ),
          onChangedMock(mockConnectionId, null),
          setArmingMock(mockConnectionId, true, false, setDisarmed),
        ]}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());

    // expect the button to look correct
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(getByTestId("connect-button"));

    await waitFor(() => expect(connectionState(cache).connecting).toBeTruthy());
    expect(getByTestId("disconnect-button")).toBeVisible();

    // expect the button to have connecting text
    expect(asFragment()).toMatchSnapshot();

    await waitFor(() => expect(connect).toHaveBeenCalled());
    await waitFor(() =>
      expect(connectionState(cache).connection).toEqual(mockConnectionId)
    );
    await waitFor(() => expect(setDisarmed).toHaveBeenCalled());
    expect(connectionState(cache).connecting).toBeFalsy();

    // expect disconnect button
    expect(getByTestId("disconnect-button")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();

    // The correct things should have been logged
    expect(logs(cache)).toMatchSnapshot();
  });

  it("should close the connection when the disconnect button is clicked", async () => {
    const mockConnectionId = "435345";
    const disconnect = jest.fn();

    const cache = clientCache();
    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: mockConnectionId,
          port: "/dev/something",
          baudRate: 115200,
        })}
        mocks={[
          onChangedMock(mockConnectionId, null),
          disconnectMock(mockConnectionId, disconnect),
        ]}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("disconnect-button")).toBeVisible());
    fireEvent.click(getByTestId("disconnect-button"));

    await waitFor(() => expect(disconnect).toHaveBeenCalled());
    expect(connectionState(cache).connection).toBeNull();
    expect(connectionState(cache).connecting).toBeFalsy();

    expect(logs(cache)).toMatchSnapshot();
  });

  it("should remove the connection when onConnectionChanged event occurs for the active connection with no new connectionId", async () => {
    const mockConnectionId = "238478234";
    const cache = clientCache();
    const mockSubscriptions = new MockSubscriptionLink();
    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: mockConnectionId,
          port: "/dev/something",
          baudRate: 115200,
        })}
        link={mockSubscriptions}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("disconnect-button")).toBeVisible());
    mockSubscriptions.simulateResult({
      result: {
        data: {
          onConnectionChanged: null,
        },
      },
    });
    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    expect(connectionState(cache).connection).toBeNull();
    expect(logs(cache)).toMatchSnapshot();
  });

  it("should update the connectionId when onConnectionChanged event occurs for the active connection with a new connectionId", async () => {
    const mockConnectionId = "238478234";
    const cache = clientCache();
    const mockSubscriptions = new MockSubscriptionLink();
    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: mockConnectionId,
          port: "/dev/something",
          baudRate: 115200,
        })}
        link={mockSubscriptions}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("disconnect-button")).toBeVisible());
    mockSubscriptions.simulateResult({
      result: {
        data: {
          onConnectionChanged: "new-id",
        },
      },
    });
    await waitFor(() =>
      expect(connectionState(cache).connection).toBe("new-id")
    );
    expect(logs(cache)).toMatchSnapshot();
  });

  it("should remove the active connection if onConnectionChanged event subscription fails", async () => {
    const mockConnectionId = "238478234";
    const cache = clientCache();
    const mockSubscriptions = new MockSubscriptionLink();
    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: mockConnectionId,
          port: "/dev/something",
          baudRate: 115200,
        })}
        link={mockSubscriptions}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("disconnect-button")).toBeVisible());
    mockSubscriptions.simulateResult({
      result: {
        errors: [new GraphQLError("Some error")],
      },
    });
    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    expect(connectionState(cache).connection).toBeNull();
    expect(logs(cache)).toMatchSnapshot();
  });

  it("should handle failing to connect", async () => {
    const cache = clientCache();
    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: null,
          port: "/dev/something",
          baudRate: 5050,
        })}
        mocks={[connectErrorMock("/dev/something", 5050)]}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    fireEvent.click(getByTestId("connect-button"));

    await waitFor(() => expect(connectionState(cache).connecting).toBeTruthy());

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    expect(connectionState(cache).connecting).toBeFalsy();
    expect(connectionState(cache).connection).toBeNull();
    expect(logs(cache)).toMatchSnapshot();
  });

  it("should close the connection if api version is less than minimum version", async () => {
    const connect = jest.fn();
    const disconnect = jest.fn();
    const mockConnectionId = "23432423";

    const cache = clientCache();
    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: null,
          port: "/dev/something",
          baudRate: 5050,
        })}
        mocks={[
          connectMock(
            "/dev/something",
            5050,
            mockConnectionId,
            "1.0.1",
            connect
          ),
          disconnectMock(mockConnectionId, disconnect),
        ]}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    fireEvent.click(getByTestId("connect-button"));

    await waitFor(() => expect(connectionState(cache).connecting).toBeTruthy());
    await waitFor(() => expect(connect).toHaveBeenCalled());

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    expect(disconnect).toHaveBeenCalled();
    expect(connectionState(cache).connecting).toBeFalsy();
    expect(connectionState(cache).connection).toBeNull();
    expect(logs(cache)).toMatchSnapshot();
  });

  it("should handle aborting connection attempts", async () => {
    const disconnect = jest.fn();
    const setDisarmed = jest.fn();
    const mockConnectionId1 = "435543";
    const mockConnectionId2 = "243455";

    const cache = clientCache();

    const { getByTestId } = render(
      <MockedProvider
        cache={cache}
        resolvers={resolvers({
          connecting: false,
          connection: null,
          port: "/dev/something",
          baudRate: 115200,
        })}
        key={1}
        mocks={[
          connectMock("/dev/something", 115200, mockConnectionId1, "1.40.1"),
          disconnectMock(mockConnectionId1, disconnect),
          connectMock("/dev/something", 115200, mockConnectionId2, "1.40.1"),
          onChangedMock(mockConnectionId2, null),
          setArmingMock(mockConnectionId2, true, false, setDisarmed),
        ]}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    fireEvent.click(getByTestId("connect-button"));

    fireEvent.click(await waitFor(() => getByTestId("disconnect-button")));
    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    await waitFor(() => expect(disconnect).toHaveBeenCalled());

    await waitFor(() => expect(getByTestId("connect-button")).toBeVisible());
    fireEvent.click(getByTestId("connect-button"));

    await waitFor(() => expect(getByTestId("disconnect-button")).toBeVisible());
    await waitFor(() =>
      expect(connectionState(cache).connection).toEqual(mockConnectionId2)
    );
    await waitFor(() => expect(setDisarmed).toHaveBeenCalled());
    expect(connectionState(cache).connecting).toBeFalsy();

    // The correct things should have been logged
    expect(logs(cache)).toMatchSnapshot();
  });
});
