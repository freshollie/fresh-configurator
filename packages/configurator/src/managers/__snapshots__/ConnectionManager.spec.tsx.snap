// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionManager should close the connection if api version is less than minimum version 1`] = `
Array [
  "Opening connection on /dev/something",
  "Serial port <span class=\\"message-positive\\">successfully</span> opened, connectionId=23432423",
  "MultiWii API version: <strong>1.0.1</strong>",
  "MSP version not supported: <span class=\\"message-negative\\">1.0.1</span>",
  "Serial port <span class=\\"message-positive\\">successfully</span> closed for connectionId=23432423",
]
`;

exports[`ConnectionManager should close the connection when the disconnect button is clicked 1`] = `
Array [
  "Serial port <span class=\\"message-positive\\">successfully</span> closed for connectionId=435345",
]
`;

exports[`ConnectionManager should connect with the connection settings and set the new connection id when connect clicked 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 70px;
  width: 70px;
  color: #fff;
  font-size: 12px;
  font-weight: normal;
  text-shadow: 0px 1px rgba(0,0,0,0.25);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #ffbb00;
  border: 1px solid #dba718;
  height: 50px;
  width: 50px;
  border-radius: 100px;
  padding-top: 0px;
  padding-right: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  cursor: pointer;
}

.c1 svg {
  width: 44px;
  height: 44px;
}

.c1:hover {
  background-color: #ffcc3f;
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="connect-button"
      type="button"
    >
      <svg
        data-name="usb-connect"
      />
    </button>
    <div>
      Connect
    </div>
  </div>
</DocumentFragment>
`;

exports[`ConnectionManager should connect with the connection settings and set the new connection id when connect clicked 2`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 70px;
  width: 70px;
  color: #fff;
  font-size: 12px;
  font-weight: normal;
  text-shadow: 0px 1px rgba(0,0,0,0.25);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #ffbb00;
  border: 1px solid #dba718;
  height: 50px;
  width: 50px;
  border-radius: 100px;
  padding-top: 0px;
  padding-right: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  cursor: pointer;
}

.c1 svg {
  width: 44px;
  height: 44px;
}

.c1:hover {
  background-color: #ffcc3f;
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="disconnect-button"
      type="button"
    >
      <svg
        data-name="usb-disconnect"
      />
    </button>
    <div>
      Connecting
    </div>
  </div>
</DocumentFragment>
`;

exports[`ConnectionManager should connect with the connection settings and set the new connection id when connect clicked 3`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 70px;
  width: 70px;
  color: #fff;
  font-size: 12px;
  font-weight: normal;
  text-shadow: 0px 1px rgba(0,0,0,0.25);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #ffbb00;
  border: 1px solid #dba718;
  height: 50px;
  width: 50px;
  border-radius: 100px;
  padding-top: 0px;
  padding-right: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  cursor: pointer;
  background-color: #e60000;
  border: 1px solid #fe0000;
}

.c1 svg {
  width: 44px;
  height: 44px;
}

.c1:hover {
  background-color: #ffcc3f;
  background-color: #f21212;
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="disconnect-button"
      type="button"
    >
      <svg
        data-name="usb-disconnect"
      />
    </button>
    <div>
      Disconnect
    </div>
  </div>
</DocumentFragment>
`;

exports[`ConnectionManager should connect with the connection settings and set the new connection id when connect clicked 4`] = `
Array [
  "Opening connection on /dev/something",
  "Serial port <span class=\\"message-positive\\">successfully</span> opened, connectionId=23455",
  "MultiWii API version: <strong>1.40.1</strong>",
  "<b>Arming disabled</b>",
]
`;

exports[`ConnectionManager should handle aborting connection attempts 1`] = `
Array [
  "Opening connection on /dev/something",
  "Connection attempt to /dev/something aborted",
  "Opening connection on /dev/something",
  "Serial port <span class=\\"message-positive\\">successfully</span> opened, connectionId=243455",
  "MultiWii API version: <strong>1.40.1</strong>",
  "<b>Arming disabled</b>",
]
`;

exports[`ConnectionManager should handle failing to connect 1`] = `
Array [
  "Opening connection on /dev/something",
  "Could not open connection (<span class=\\"message-negative\\">some error</span>), communication <span class=\\"message-negative\\">failed</span>",
]
`;

exports[`ConnectionManager should remove the active connection if onConnectionChanged event subscription fails 1`] = `
Array [
  "Serial port closed unexpectedly for connectionId=238478234",
]
`;

exports[`ConnectionManager should remove the connection when onConnectionChanged event occurs for the active connection with no new connectionId 1`] = `
Array [
  "Serial port closed unexpectedly for connectionId=238478234",
]
`;

exports[`ConnectionManager should update the connectionId when onConnectionChanged event occurs for the active connection with a new connectionId 1`] = `
Array [
  "Serial port reconnected for connectionId=238478234 newConnectionId=new-id ",
]
`;
